documentation/modules/managing/con-consumer-config-properties.adoc

```diff
@@ -0,0 +1,253 @@
+If this impacts throughput unduly, you can commit less frequently,
+or you can use the `commitAsync` API.
+The `commitAsync` API does not wait for the broker to respond to a commit request,
+but risks creating more duplicates when rebalancing.
+A common approach is to combine both commit APIs in an application, with the `commitSync` API used just before shutting the consumer down or rebalancing to make sure the final commit is successful.
+
+Also, consider using transactional ids and having idempotence enabled (`enable.idempotence=true`) on the producer side to guarantee exactly once delivery.
+On the consumer side, you can then use the `isolation.level` property to control how transactional messages are read by the consumer.
+
+The `isolation.level` property has two valid values:
+
+* `read_committed`
+* `read_uncommitted`
+
+Use `read_committed` to ensure that only transactional messages that have been committed are read by the consumer.
+However, this will cause an increase in end-to-end latency, because the consumer won’t be able to return a message until the brokers have written the transaction markers that record the result of the transaction (_committed_ or _aborted_).
+
+----
+# ...
+enable.auto.commit=false <1>
+isolation.level=read_committed <2>
+# ...
+----
+<1> Auto commit is set to false to provide more control over committing offsets
+<2> Set to `read_committed` so that only committed messages are read by the consumer.
+
+== Recovering from failure to avoid data loss
+
+Use the `session.timeout.ms` and `heatbeat.interval.ms` properties to configure the time taken to check and recover from consumer failure within a consumer group.
+
+The `session.timeout.ms` property specifies the maximum amount of time in milliseconds a consumer within a consumer group can be out of contact with a broker before being considered inactive and a _rebalancing_ is triggered between the active consumers in the group.
+When the group rebalances, the partitions are reassigned to the members of the group.
+
+The `heatbeat.interval.ms` property specifies the interval in milliseconds between _heartbeat_ checks to the consumer group coordinator to indicate that the consumer is active and connected.
+The heartbeat interval must be lower, usually by a third, than the session timeout interval.
+
+If you set the `session.timeout.ms` property lower, failing consumers are detected earlier, and rebalancing can take place quicker.
+However, take care not to set the timeout so low that the broker fails to receive a heartbeat in time and triggers an unnecessary rebalance.
+
+Decreasing the heartbeat interval reduces the chance of accidental rebalancing, but more frequent heartbeats increases the overhead on broker resources.
+
+== Handling uncommitted offsets
+
+Use the `auto.offset.reset` property to control how the consumer behaves when offsets are not committed.
```

Comment:
The title and the first sentence seem misleading. I expected this to talk about what to do when the commit call fails. But this is rather what to do when there are no previous offsets. Maybe you could make it a bit clearer?

Reply:
Yes. Could be clearer. I've changed to:

> Use the `auto.offset.reset` property to control how a consumer behaves when no offsets have been committed,
> or a committed offset is no longer valid or deleted.
> 
> Suppose a consumer fails before an offset has been committed. The consumer that then assumes responsibility for consuming the messages will start reading from the reassigned topic partition based on the reset policy.
> To avoid data loss, but increase the amount of processing, use the `earliest` value to start at the beginning of the partition. The default value is `latest`, which starts at the end of the partition, and consequently means some messages are missed.